filetracker/cachecleaner.py

```diff
@@ -0,0 +1,254 @@
+        """
+        logger.info("Starting daemon.")
+        while True:
+            try:
+                self._scan_disk()
+                do_cleaning, delete_from_index = self._analyze_file_index()
+                if do_cleaning:
+                    self._clean_cache(delete_from_index)
+            except Exception:
+                logger.exception("Following exception occurred:")
+            sleeping_until_time = datetime.datetime.now() + self.scan_interval
+            logger.info("Sleeping until %s.", sleeping_until_time)
+            time.sleep(self.scan_interval.total_seconds())
+
+    def _scan_disk(self):
+        logger.info("Scanning disk...")
+        dirs = []
+        for glob_cache_dir in self.glob_cache_dirs:
+            dirs.extend(glob.glob(glob_cache_dir))
+        logger.debug("Glob expressions have expanded to %s", dirs)
+        self.clients = [filetracker.Client(cache_dir=dir, remote_store=None)
+                        for dir in dirs]
+        self.file_index = []
+        for client in self.clients:
+            for file_info in client.list_local_files():
+                self.file_index.append(
+                    CacheCleaner.FileIndexEntry(file_info=file_info,
+                                                client=client))
+
+    def _analyze_file_index(self):
+        logger.info("Analyzing cache...")
+
+        self.file_index.sort(key=lambda x: (-x.file_info.mtime,
+                                            x.file_info.size))
+        cache_size = 0
+        delete_from_index = None
+        do_cleaning = False
+        for i, entry in enumerate(self.file_index):
+            cache_size += entry.file_info.size
+            if (delete_from_index is None and
+                    cache_size >= self.cleaning_level):
+                delete_from_index = i
+            if cache_size >= self.cache_size_limit:
+                do_cleaning = True
+
+        logger.info("Analysis done. Cache size: %s.",
+                    format_size_with_unit(cache_size))
+        if not do_cleaning:
+            logger.info("Decided not to perform cache cleaning.")
+
+        return do_cleaning, delete_from_index
+
+    def _clean_cache(self, delete_from_index):
+        # assumption: run after cache analyze
+        logger.info("Performing cache cleaning...")
+        deleted_files_cnt = 0
+        deleted_bytes = 0
+        assert (self.file_index[0].file_info.mtime
+                >= self.file_index[-1].file_info.mtime)
+        for entry in self.file_index[delete_from_index:]:
+            logger.debug("Deleting file: %s from store located at: %s",
+                         entry.file_info.name, entry.client.local_store.dir)
+            try:
+                entry.client.delete_file(entry.file_info.name)
+                deleted_files_cnt += 1
+                deleted_bytes += entry.file_info.size
+            except Exception as e:
+                logger.debug(e, exc_info=True)
```

Comment:
LocalDataStore wyjątkiem nie rzuci, bo Client go złapie, więc faktycznie jedyny problem jaki może być, to z lockami. W takim razie niedługo to poprawię tak jak zasugerowałeś.